aw_client/client.py

```diff
@@ -72,72 +60,155 @@ def _get(self, endpoint: str) -> Optional[req.Response]:
+        if self.connected:
+            try:
+                self._post(endpoint, data)
+                logger.debug("Sent event to server: {}".format(event))
+            except req.RequestException as e:
+                self._connection_lost()
+                logger.warning("Failed to send event to server ({})".format(e))
+                self._queue_failed_request(endpoint, data)
+        else:
+            self._queue_failed_request(endpoint, data)
+
+    #
+    #   Bucket get/post requests
+    #
 
     def get_buckets(self):
         return self._get('buckets').json()
 
-    def get_events(self, bucket) -> List[Event]:
-        endpoint = "buckets/{}/events".format(bucket)
-        events = self._get(endpoint).json()
-        return [Event(**event) for event in events]
+    def setup_bucket(self, bucket_id, event_type: str) -> bool:
+        self.buckets.append({"bid": bucket_id, "etype": event_type})
+        if self.connected:
+            self._create_buckets()
 
-    def create_bucket(self, bucket_id, event_type: str) -> bool:
+    def _create_buckets(self):
         # Check if bucket exists
         buckets = self.get_buckets()
-        if bucket_id in buckets:
-            return False  # Don't do anything if bucket already exists
+        for bucket in self.buckets:
+            if bucket['bid'] in buckets:
+                return False  # Don't do anything if bucket already exists
+            else:
+                # Create bucket
+                endpoint = "buckets/{}".format(bucket['bid'])
+                data = {
+                    'client': self.client_name,
+                    'hostname': self.client_hostname,
+                    'type': bucket['etype'],
+                }
+                try:
+                    self._post(endpoint, data)
+                except req.RequestException as e:
+                    logger.error("Failed to create bucket: {}".format(e))
+                return True
+
+    #
+    #   Failed request queue handling
+    #
+
+    def _queue_failed_request(self, endpoint: str, data: dict):
+        # Find failed queue file
+        entry = {"endpoint": endpoint, "data": data}
+        self.queue_file_lock.acquire()
+        with open(self.queue_file, "a+") as queue_fp:
+            queue_fp.write(json.dumps(entry) + "\n")
+        self.queue_file_lock.release()
+
+    def _post_failed_requests(self):
+        failed_requests = []
+        self.queue_file_lock.acquire()
+        with open(self.queue_file, "r") as queue_fp:
+            for request in queue_fp:
+                failed_requests.append(json.loads(request))
+            if len(failed_requests) != 0:
+                open(self.queue_file, "w").close()  # Clear file
+                logger.info("Sending {} failed events: {}".format(len(failed_requests), failed_requests))
+                for request in failed_requests:
+                    self._post(request['endpoint'], request['data'])
```

Comment:
It was a lot easier on the earlier implementation to fix this issue because it only handled events instead of requests and we could send a giant request will the whole queue. To support replace_last_event i had to make them into requests instead.

A alternative that could work would be to always have one event in a buffer before sending it, so replace_last is always a local thing on the aw-client, but the reason why we didn't go with that implementation from the beginning was in case someone would be watching a movie for 2hrs and then they had a power outage, crash or whatever that very significant event would be lost.